iceoryx_hoofs/test/moduletests/test_posix_semaphore_interface.cpp

```diff
@@ -0,0 +1,264 @@
+    constexpr iox::units::Duration timeToWait = 2_ms;
+
+    for (uint64_t i = 0; i < NUMBER_OF_INCREMENTS; ++i)
+    {
+        ASSERT_FALSE(this->sut->post().has_error());
+    }
+
+    for (uint64_t i = 0; i < NUMBER_OF_DECREMENTS; ++i)
+    {
+        auto call = this->sut->timedWait(timeToWait);
+        ASSERT_FALSE(call.has_error());
+        ASSERT_TRUE(call.value() == iox::posix::SemaphoreWaitState::NO_TIMEOUT);
+    }
+
+    auto result = this->sut->getState();
+    ASSERT_THAT(result.has_error(), Eq(false));
+    EXPECT_THAT(result->value, Eq(NUMBER_OF_INCREMENTS - NUMBER_OF_DECREMENTS));
+}
+
+TYPED_TEST(SemaphoreInterfaceTest, FailingTimedWaitDoesNotChangeSemaphoreValue)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "cb2f5ded-7c04-4e2d-ab41-bbe231a520c7");
+    constexpr uint64_t NUMBER_OF_DECREMENTS = 4U;
+    constexpr iox::units::Duration timeToWait = 2_us;
+
+    for (uint64_t i = 0; i < NUMBER_OF_DECREMENTS; ++i)
+    {
+        auto call = this->sut->timedWait(timeToWait);
+        ASSERT_FALSE(call.has_error());
+        ASSERT_TRUE(call.value() == iox::posix::SemaphoreWaitState::TIMEOUT);
+    }
+
+    auto result = this->sut->getState();
+    ASSERT_THAT(result.has_error(), Eq(false));
+    EXPECT_THAT(result->value, Eq(0));
+}
+
+
+TYPED_TEST(SemaphoreInterfaceTest, TryWaitAfterPostIsSuccessful)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "df5d32f5-9417-419e-b36e-93a0502ed6e7");
+    ASSERT_FALSE(this->sut->post().has_error());
+    auto call = this->sut->tryWait();
+    ASSERT_THAT(call.has_error(), Eq(false));
+    ASSERT_THAT(*call, Eq(true));
+}
+
+TYPED_TEST(SemaphoreInterfaceTest, TryWaitWithNoPostIsNotSuccessful)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "8d0eebcd-11e3-456f-8d19-3fdb6fc70241");
+    ASSERT_FALSE(this->sut->post().has_error());
+    auto call = this->sut->tryWait();
+    ASSERT_THAT(call.has_error(), Eq(false));
+    ASSERT_THAT(*call, Eq(true));
+}
+
+TYPED_TEST(SemaphoreInterfaceTest, WaitValidAfterPostIsNonBlocking)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "c08220e3-c368-43d5-85d2-cbf7e1659017");
+    ASSERT_FALSE(this->sut->post().has_error());
+    // this call should not block and should be successful
+    EXPECT_THAT(this->sut->wait().has_error(), Eq(false));
+}
+
+TYPED_TEST(SemaphoreInterfaceTest, WaitBlocksAtLeastDefinedSleepTime)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "9fb454db-2d1d-43a5-b0a5-d9de7f1886e0");
+
+    std::chrono::steady_clock::time_point start;
+    std::thread t1([&] {
+        start = std::chrono::steady_clock::now();
+        std::this_thread::sleep_for(std::chrono::nanoseconds(this->TIMING_TEST_WAIT_TIME.toNanoseconds()));
+        ASSERT_FALSE(this->sut->post().has_error());
+    });
+
+
+    ASSERT_FALSE(this->sut->wait().has_error());
+    auto end = std::chrono::steady_clock::now();
+
+    EXPECT_THAT(std::chrono::nanoseconds(end - start).count(), Ge(this->TIMING_TEST_WAIT_TIME.toNanoseconds()));
+
+    t1.join();
```

Comment:
Please move `join` before the line with the access to `start` in order to sync the memory